core/src/main/java/com/linecorp/armeria/common/stream/AsyncMapStreamMessage.java

```diff
@@ -0,0 +1,218 @@
+/*
+ * Copyright 2021 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.stream;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.internal.common.stream.StreamMessageUtil;
+
+import io.netty.util.concurrent.EventExecutor;
+
+final class AsyncMapStreamMessage<T, U> implements StreamMessage<U> {
+    private final StreamMessage<Object> source;
+    private final Function<Object, CompletableFuture<U>> function;
+
+    @SuppressWarnings("unchecked")
+    AsyncMapStreamMessage(StreamMessage<? extends T> source,
+                          Function<? super T, ? extends CompletableFuture<? extends U>> function) {
+        requireNonNull(source, "source");
+        requireNonNull(function, "function");
+
+        this.source = (StreamMessage<Object>) source;
+        this.function = (Function<Object, CompletableFuture<U>>) function;
```

Comment:
We don't need to convert `StreamMessage<? extends T> source` to `StreamMessage<Object>` unlike `FuseableStreamMessage`.
Could we make this implementation more type-safe?